api/envoy/service/ext_proc/v3alpha/external_processor.proto

```diff
@@ -0,0 +1,307 @@
+  // filter. This metadata will be placed in the namespace "envoy.filters.http.ext_proc".
+  google.protobuf.Struct dynamic_metadata = 8;
+
+  // Override how parts of the HTTP request and response are processed
+  // for the duration of this particular request/response only. Servers
+  // may use this to intelligently control how requests are processed
+  // based on the headers and other metadata that they see.
+  envoy.extensions.filters.http.ext_proc.v3alpha.ProcessingMode mode_override = 9;
+}
+
+// The following are messages that are sent to the server.
+
+// This message is sent to the external server when the HTTP request and responses
+// are first received.
+message HttpHeaders {
+  // The HTTP request headers. All header keys will be
+  // lower-cased, because HTTP header keys are case-insensitive.
+  config.core.v3.HeaderMap headers = 1;
+
+  // The values of properties selected by the "request_properties"
+  // list in the configuration. Each entry in the map is the
+  // value of the specified property.
+  map<string, google.protobuf.Struct> properties = 2;
+
+  // If true, then there is no message body associated with this
+  // request or response.
+  bool end_of_stream = 3;
+}
+
+// This message contains the message body that Envoy sends to the external server.
+message HttpBody {
+  bytes body = 1;
+
+  bool end_of_stream = 2;
+}
+
+// This message contains the trailers.
+message HttpTrailers {
+  config.core.v3.HeaderMap trailers = 1;
+}
+
+// The following are messages that may be sent back by the server.
+
+// This message must be sent in response to an HttpHeaders message.
+message HeadersResponse {
+  CommonResponse response = 1;
+}
+
+// This message must be sent in response to an HttpTrailers message.
+message TrailersResponse {
+  // Instructions on how to manipulate the trailers
+  HeaderMutation header_mutation = 1;
+}
+
+// This message must be sent in response to an HttpBody message.
+message BodyResponse {
+  CommonResponse response = 1;
+}
+
+// This message contains common fields between header and body responses.
+// [#next-free-field: 6]
+message CommonResponse {
+  enum ResponseStatus {
+    // Apply the mutation instructions in this message to the
+    // request or response, and then continue processing the filter
+    // stream as normal. This is the default.
+    CONTINUE = 0;
+
+    // Replace the request or response with the contents
+    // of this message. If header_mutation is set, apply it to the
+    // headers. If body_mutation is set and contains a body, then add that
+    // body to the request or response, even if one does not already exist --
+    // otherwise, clear the body. Any additional body and trailers
+    // received from downstream or upstream will be ignored.
+    // This can be used to add a body to a request or response that does not
+    // have one already.
+    CONTINUE_AND_DO_NOT_END_STREAM = 1;
```

Comment:
I'm having a hard time parsing this name in contrast to `CONTINUE` above. Does the above `CONTINUE` end the stream and this one doesn't? This seems a little odd.

Reply:
The intention of this filter was to support the same kind of workflow as the ContinueAndDontEndStream status in FilterHeadersStatus, which is used to add a body and trailers to an HTTP request that does not have them. So I copied the name. But if you have an idea for a better name then that'd be great!

Reply:
Envoys internal naming isn't great as it's not a stable public API, I think in this case we shoul

How about `REPLACE`?

What does Wasm do here @PiotrSikora ?

Reply:
I'm changing it to CONTINUE_AND_REPLACE, which makes it clear that we're still continuing processing.

Reply:
We don't have this case covered in Wasm right now, but to be honest, I'm not a big fan of mixing control of the processing flow (stop/continue) with mutating `end_stream` value.

Reply:
The original intention of this back and forth was to allow a scenario in which we:
1) Tell the filter to continue but not end the stream.
2) Later send more messages to eventually complete the stream.

At it has been updated and now written, I agree that CONTINUE_AND_REPLACE is what makes sense, but we should make it very clear this is basically END_STREAM_AND_CONTINUE or CONTINUE_AND_END_STREAM (would that be better?)

In the future we can add CONTINUE_AND_DONT_END_STREAM

Reply:
The issue is that we have 3 options here:

1) `Continue` (and don't modify `end_stream`),
2) `ContinueAndEndStream` (it was removed last month in #13216, but the justification seems wrong),
3) `ContinueAndDontEndStream`.

and we really should have the same values available in `FilterDataStatus`, since we might want to truncate body and/or replace it with something else, which ends up with a weird interface where request/response body is mutated one way, and the `end_stream` flag for the same request/response is mutated in a different way.

I think it would be useful if we could decouple those 2 and limit use of return values to control processing flow (i.e. continue/stop), and mutate request/response body or network bytestream via APIs.

Reply:
I agree, and these are the cases I was pushing for when we were iterating on this PR. I agree the C++ API is suboptimal, but it is what it is and has evolved over time.

I want to make sure we can support all of these cases in both WASM and with ext_proc so IMO we should focus more on making sure we can support these cases eventually, even if we don't do it now, and worry less about the C++ implementation.